### ISSUE TITLE
`/run/wrappers` not mounted when `boot.initrd.systemd.enable = true`

### ISSUE THREAD

**Issue by someplaceguy** [NixOS/nixpkgs#244737]
### Describe the bug

When I set `boot.initrd.system.enable = true`, `/run/wrappers` doesn't get mounted nor does it get populated with the suid wrappers.

This makes `su`, `sudo` and `doas` unusable.

I'm not sure if anything else from the activation script is not working / running either.

### Steps To Reproduce

All I did was enable the mentioned systemd in initrd option on a system running NixOS 23.05, run `nixos-rebuild boot` and then reboot.

### Expected behavior

I expected `/run/wrappers` to be mounted, to have a `bin` symlink and for `sudo`, `su` and `doas` to work.

### Additional context

My NixOS configuration is extensive/complex and I do have local modifications in `nixpkgs`, so it's not inconceivable that it is related to something I'm not doing right.

### Notify maintainers

cc @ElvishJerricco 

Not sure if I should cc someone else, perhaps @ElvishJerricco knows.

**Comment by ElvishJerricco** [comment]
That's very odd. I certainly have `/run/wrappers/bin` on all my systems with systemd-initrd. One thing to note is that we move `postBootCommands` and activation scripts so they run in a chroot during stage 1, rather than running just before systemd in stage 2. The reason for this is so that `systemctl switch-root` can pass systemd state from stage 1 to stage 2, but it requires stage 2 init to be the actual systemd binary instead of a wrapper script like it is without systemd-initrd.

Point being, if you're using activation scripts or `boot.postBootCommands` (which (effectively) includes `powerManagement.powerUpCommands`) to do anything fancy, those may be experiencing slight confusion for running in a chroot? We tried to make it compatible though.

**Comment by someplaceguy** [comment]
On this system, I am not using `boot.postBootCommands` or `powerManagement.powerUpCommands` but I was using this activation script:

```Nix
   activationScripts = {
      termpoweroff = lib.stringAfter [ "etc" "specialfs" "var" "usrbinenv" "udevd" "hostname" "domain" "nix" "users" "groups" "binsh" "stdio" ] ''
        for i in $(seq 0 256); do
          ${pkgs.utillinux}/bin/setterm --blank=5 --powerdown=1 --term linux >> /dev/tty$i
        done
      '';
    };
```

I tried disabling this activation script and it sems to have solved the `/run/wrappers` mounting issue but it somehow introduced different, perhaps even more serious problems when booting:

![IMG_20230722_195541](https://github.com/NixOS/nixpkgs/assets/138138196/b5912908-dff3-4731-ad09-6f305d8063ff)

Apparently, I couldn't even login as root.

It's quite baffling how removing the activation script can cause these problems, so I'm now wondering if it was just a coincidence and in fact it's an unrelated problem.

But when I pressed <Enter>, the boot process seemed to continue fine.

Unfortunately, I am not using `journald` and the error messages printed during boot, if any, didn't make it to the logs, apparently because they weren't printed to `/dev/kmsg`. So I guess I should figure out how to enable rsyslogd as a stage1 logging daemon (assuming it is even possible).

**Comment by someplaceguy** [comment]
Ok, so I have booted this configuration (without the activation script) two more times, and it seems that the activation script was a red herring: I am back to the same issue as in the initial description.

So apparently the activation script that I had has nothing to do with the `/run/wrappers` mounting issue. 

I'm guessing this also means that the boot failure I experienced before is probabilistic and apparently, not easy to reproduce given that I only experienced this failure once in about 10 boots total that I've done so far with systemd in initrd.

By the way, even though `/run/wrappers` is not mounted automatically, I am able to manually fix the issue by logging in as root and then running `/run/current-system/activate`, but of course, this is not really what a user should have to do.

**Comment by someplaceguy** [comment]
I'm attaching a log of my boot process, but **please note that the timestamps are highly misleading**, i.e. the order of events should be correct but the time between events is not:

[log2.txt](https://github.com/NixOS/nixpkgs/files/12140028/log2.txt)

When I compare it the log in #245089, in which I had the same configuration but `/run/wrappers` actually did mount successfully, here's what jumps out the most for me:

1. In the boot where `/run/wrappers` was unmounted, the `NixOS Activation` service runs **before** `/sysroot/run/wrappers` gets unmounted.
2. In the boot where '/run/wrappers` was mounted properly, the `NixOS Activation` service runs **after** `/sysroot/run/wrappers` gets unmounted.

I'm not sure if this is the root cause, as I am out of my depth here, but maybe it helps.

**Comment by ElvishJerricco** [comment]
@someplaceguy I believe the timestamps would be right if you used journald instead of rsyslogd.

**Comment by someplaceguy** [comment]
> @someplaceguy I believe the timestamps would be right if you used journald instead of rsyslogd.

Sure, and I really like journald's features and admin interface, but when I used it a few years ago, its implementation was extremely unreliable, so I simply cannot use it.

In other words, I really need my log daemon to be reliable if I want to have any hope of diagnosing hard-to-debug issues like the ones we've been discussing or many others that I run into over the years.

Here are some bugs in journald that (at least a few years ago) frequently caused me to lose a large amount of log messages:

1. journald was not able to properly handle jumps in the system clock.

This was especially problematic in my Raspberry Pis, which don't have a battery-backed clock and therefore always booted up with the time set to the year 1970 (or whatever it is), until `chronyd` has a chance to connect to an NTP server and set the clock correctly.

This would actually lead me to losing entire days worth of log messages (in my Raspberry Pis, at least), for no good reason. 

This also caused many bugs in the `journalctl` utility, as it wasn't able to track different boots correctly and many other similar issues.

systemd also has had some bugs with jumps in the system clock, but they are/were far less severe -- I think the worst that happens is that some timer units are started when they clearly shouldn't be.

2. journald lost log messages when a service was starting, immediately sent some messages to stderr and then immediately failed.

This made debugging some service failures much harder, as it is not even clear to the user when this happens -- the error messages were simply lost and there was no way to retrieve them nor was there any indication that these messages were lost.

If I recall correctly, the journald developers even said they weren't going to fix this issue, although I could be remembering incorrectly (this was a few years ago).

3. journald was not able to handle power loss without corrupting its log files (?)

I don't necessarily mind that journald uses a binary format to store its logs, although I wish there was a simple tool to read them, e.g. when booted from a rescue system (perhaps journald can do that?).

But to me it's unacceptable that the logs corrupt themselves just because of an abrupt power down.

To be clear, this may actually not be an issue with journald's design: it's impossible to distinguish between an abrupt power loss and a maliciously (or accidentally) truncated log file, and it's useful to know when this happens.

However, even if that's the case and hopefully (!!!) no messages are being lost unnecessarily (due to journald being unable to fully parse a truncated log correctly), the user interface should definitely be improved: there should be a clear distinction between a log that was simply truncated and a log that was corrupted in any other way, as the latter is a much more serious issue than the former.

There was also another unrelated problem that led to journald corrupting its log files very frequently on some of my machines, but I think this may have been a bug in ZFS's mmap implementation that was fixed recently. I only experienced this corruption with journald, even when there were no abrupt power downs, but I think it's because journald made heavy use of mmap for its log files in strange ways instead of doing I/O normally (although this is also a red flag for me, as I think mmap precludes handling I/O failures gracefully, at least in some cases).

In short, I wouldn't mind using journald but they'd definitely have to fix the above issues first, especially those that lead to message loss.

**Comment by ElvishJerricco** [comment]
We're not here to debate the pros and cons of journald in all contexts. I'm not asking you to switch your systems to journald. I'm telling you we'd get better logs in this particular bug report if journald were used. That matters a lot and right now it's very hard to help with how little info we have, including the fact that you haven't shared almost any of your configuration.

**Comment by someplaceguy** [comment]
> I'm not asking you to switch your systems to journald. 

@ElvishJerricco Sorry, I misinterpreted you. I thought you were suggesting me to switch to journald.

> That matters a lot and right now it's very hard to help with how little info we have, including the fact that you haven't shared almost any of your configuration.

The NixOS configuration for this system (my laptop) is 2 thousand lines of code, plus I have about 200 patches applied to `nixpkgs`, so I'm not sure sharing all of that information would necessarily be productive nor would I like to do that, for privacy reasons.

If you'd like me to provide some specific information or at least narrow it down to a few specific options or even an option tree, or run any commands, or change anything, I could try to do that instead.

Or if you could provide a comment similar to the one you provided in https://github.com/NixOS/nixpkgs/issues/245089#issuecomment-1646966283 then I can try looking into those details to see if I can debug the issue myself, or at least see if I can get more relevant information. As an example, right now I am in the process of trying to extract my initrd and look into the configuration of the units you mentioned in that issue to see if anything looks suspicious.

In any case, if you don't want or you don't think you can help me this way, that's also perfectly fine! You have no obligation to help debug my issues, some of which may be unique to my configuration. 

In any case, I really appreciate and am very grateful for all the help you've provided so far! Thank you.

**Comment by ElvishJerricco** [comment]
> @ElvishJerricco Sorry, I misinterpreted you. I thought you were suggesting me to switch to journald.

I was suggesting that you try it for a couple of boots just to get the more useful logs. The ordering of systemd-related events is very very important in this context and I expect journald to order those most correctly, especially if you're saying rsyslogd can't get the timestamps right.

> The NixOS configuration for this system (my laptop) is 2 thousand lines of code

Your entire 2Kloc config isn't necessary, but the bits that have anything to do with initrd would be very important to this issue. Anything at all that affects the generation of initrd, the activation scripts, or the `postBootCommands` should pretty much cover anything we'd need to see to understand an initrd failure. Like if you have any `boot.initrd.{pre,post}{Mount,Device,LVM}Commands` style options, then consider that those don't work in systemd stage 1: https://github.com/NixOS/nixpkgs/pull/227633

**Comment by someplaceguy** [comment]
> I was suggesting that you try it for a couple of boots just to get the more useful logs. The ordering of systemd-related events is very very important in this context and I expect journald to order those most correctly, especially if you're saying rsyslogd can't get the timestamps right.

As I said before in this issue, the ordering between events is correct in my logs. These log messages are sent by journald in the correct order to `/dev/kmsg`, which `rsyslogd` then copies to `/var/log/messages.log` in the same order.

However, I configured `rsyslogd` to ignore the `/dev/kmsg` timestamps due to an unrelated issue, so the timestamp that gets applied to these boot messages is the timestamp when `rsyslog` copied each individual message from `/dev/kmsg` to `/var/log/messages.log`, not the timestamp that was saved when the message was stored in `/dev/kmsg`.

So in this case, you cannot rely on the *elapsed time between events* in my logs (for those events that happen before `rsyslogd` starts), but you should be able to rely on how the messages are ordered between themselves.

I could try using `journald` if the elapsed time between events would be useful information to have.

> Your entire 2Kloc config isn't necessary, but the bits that have anything to do with initrd would be very important to this issue. Anything at all that affects the generation of initrd, the activation scripts, or the `postBootCommands` should pretty much cover anything we'd need to see to understand an initrd failure. Like if you have any `boot.initrd.{pre,post}{Mount,Device,LVM}Commands` style options, then consider that those don't work in systemd stage 1: #227633

Ok, let's see. I don't know exactly what affects the generation of initrd, but here are all the `boot.initrd.*` options that I have set in this system:

```Nix
     initrd = {
        availableKernelModules = [ "hid_apple" "hid_microsoft" "usbhid" "usb_storage" "uas" ];
      };
```
```Nix
   initrd = {
      availableKernelModules = [ "nvme" "ehci_pci" "ahci" "xhci_pci" "sd_mod" ];
      luks.devices = {
        root = {
          device = "/dev/disk/by-uuid/f112da17-15e3-4e34-a61c-d67a661c6a0e";
          allowDiscards = true;
          bypassWorkqueues = true;
        };
      };
      systemd = {
        contents = {
          "/etc/systemd/journald.conf".text = ''
            [Journal]
            Storage=persistent
            RateLimitInterval=0
            RateLimitBurst=0
            ReadKMsg=false
            ForwardToKMsg=true
            MaxLevelKMsg=debug
            Storage=none
          '';
        };
        enable = true;
      };
    };
```

I don't have any other `boot.initrd.*` options.

But I have `boot.tmp.cleanOnBoot = true'`, not sure if it's relevant. Hopefully it's not deleting anything it shouldn't in the initrd.

In this system I believe I only have one activation script (the one I showed above), but I tried disabling it and the issue persisted.

I don't have any `boot.initrd.{pre,post}{Mount,Device,LVM}Commands`, `postBootCommands` or `powerUpCommands` in my config.

I don't think I have any patches in `nixpkgs/nixos` that could be related to these issues, except for the patches you provided in #227208 (including the `cryptsetup` dependency, which I haven't reverted yet).

... that is, unless the Network Manager services are started in initrd, which I don't think is the case?

I'm not aware of anything else in my system that could be relevant to this issue or #245089...

**Comment by ElvishJerricco** [comment]
Maybe some logs with `rd.systemd.log_level=debug` on the cmdline would help. As described so far, the ordering of systemd events makes absolutely no sense.

**Comment by someplaceguy** [comment]
Here are the requested logs:

[log3.txt](https://github.com/NixOS/nixpkgs/files/12140776/log3.txt)

Thanks!

**Comment by someplaceguy** [comment]
I've checked the logs and I've noticed the following events happen in order:

1. `initrd-nixos-activation.service` seems to run successfully. Presumably, this is when `/run/wrappers` gets mounted inside the `/sysroot` chroot.
2. Afterwards, `initrd-cleanup.service` starts
3. Then `/sysroot/run/wrappers` get unmounted, along with other mountpoints under `/sysroot`
4. Finally, the "Switch Root" service is started

So it seems that `/run/wrappers` gets unmounted just before switching the root and then the activation script does not run again, so it never gets remounted afterwards.

**Comment by ElvishJerricco** [comment]
> Presumably, /run/wrappers got unmounted due to the systemctl isolate command, which stops all the mount units that are not dependencies of the switch-root target?

If this were a problem, *no configuration with systemd initrd would work*. But in fact, `initrd-switch-root.service` has `Wants=initrd-fs.target`, which keeps FSes around upon `isolate`.

**Comment by someplaceguy** [comment]
> If this were a problem, _no configuration with systemd initrd would work_. But in fact, `initrd-switch-root.service` has `Wants=initrd-fs.target`, which keeps FSes around upon `isolate`.

But `initrd-fs.target` does not contain any dependencies on the `sysroot-run-wrappers.mount` unit.

This is what the `systemd.special` man page says:
```
      initrd-fs.target
           systemd-fstab-generator(3) automatically adds dependencies of type Before= to sysroot-usr.mount and all mount points found in /etc/fstab that have the
           x-initrd.mount mount option set and do not have the noauto mount option set.
```

So, the dependency between the `sysroot-run-wrappers.mount` unit and the `initrd-fs.target` unit is supposed to be added by `systemd-fstab-generator`.

But as you can see from the logs, `systemd-fstab-generator` does not find the `/run/wrappers` filesystem.

In fact, if you search for `Found entry` in the latest debug log that I posted, only this is found:
```
586:2023-07-24T01:16:24.597184+00:00 wizylap daemon.debug systemd-fstab-generator[116]: Found entry what=rpool where=/ type=zfs makefs=no growfs=no pcrfs=no noauto=no nofail=no
11718:2023-07-24T01:16:25.002735+00:00 wizylap daemon.debug systemd-sysroot-fstab-check[667]: Found entry what=rpool where=/ type=zfs makefs=no growfs=no pcrfs=no noauto=no nofail=no
11965:2023-07-24T01:16:25.012049+00:00 wizylap daemon.debug systemd-fstab-generator[672]: Found entry what=rpool where=/ type=zfs makefs=no growfs=no pcrfs=no noauto=no nofail=no
```

(And I think even this entry is wrong because I think it should be `where=/sysroot`, not `where=/`).

**Comment by ElvishJerricco** [comment]
While you have a good point there, it doesn't explain two things.

1. Why does no one else have this problem?
2. Why is *everything else* also being unmounted? That is positively antithetical to the dependency relationship between `initrd-switch-root.service` and `initrd-fs.target`.

**Comment by someplaceguy** [comment]
> Why does no one else have this problem?

I don't know exactly, but if you add `rd.systemd.log_level=debug` as a kernel command line parameter and post your logs, we could look and see why this doesn't happen to you.

But I can tell you that the only two times that my system did not unmount `/run/wrappers` when I enabled systemd in initrd, it was because some filesystems under `/sysroot/` could not be unmounted due to them being busy:

```
2023-07-23T20:15:03.278771+00:00 wizylap daemon.info umount[1778]: umount: /sysroot/dev: target is busy.
(...)
2023-07-23T20:15:03.278787+00:00 wizylap daemon.info umount[1779]: umount: /sysroot/run: target is busy.
(...)
```

So perhaps your boot is working correctly by accident, not because everything is configured properly.

Or perhaps your configuration is different from mine in some important way. But we won't know until we can compare our logs.

> Why is everything else also being unmounted? That is positively antithetical to the dependency relationship between initrd-switch-root.service and initrd-fs.target.

I don't understand this question. As far as I can see, they are being unmounted because `systemd-fstab-generator` cannot find any entries in either `/etc/fstab` nor `/nix/store/s43k33w1p9dqnb030w5yrwv48pjvr49x-initrd-fstab`, except for the root entry that I mentioned.

Since `systemd-fstab-generator` cannot find any entries for any filesystems, the dependencies between their mount units and `initrd-fs.target` are not being added.

This explains why the `systemctl isolate` command stops all these units even though `initrd-fs.target` is a dependency.

It does not matter that `initrd-fs.target` is a dependency because the dependencies between `initrd-fs.target` and the the mount units do not exist, which is why the mount units are getting stopped.

**Comment by someplaceguy** [comment]
The `systemd-fstab-generator.c` file contains the following comment, which provides more insight:

```C
/* If arg_sysroot_check is false, run as generator in the usual fashion.
 * If it is true, check /sysroot/etc/fstab for any units that we'd want to mount
 * in the initrd, and call daemon-reload. We will get reinvoked as a generator,
 * with /sysroot/etc/fstab available, and then we can write additional units based
 * on that file. */
```

As far as I can tell, `systemd-fstab-generator` gets called with `arg_sysroot_check = true` as part of the `initrd-parse-etc.service`.

This does not find any filesystems, and even if it did, it wouldn't actually add any dependencies between the mount units and `initrd-fs.target`, but I think that's actually what it's supposed to do.

However, it should trigger a daemon reload after `/sysroot` has been mounted, which would then reinvoke `systemd-fstab-generator` as an actual unit (and dependency) generator, which would then find the filesystem entries in `/sysroot/etc/fstab`.

In my system, I can confirm that the daemon reload is correctly triggered and that this happens after `/sysroot` has been mounted. So far so good.

However, I am not sure how `systemd-fstab-generator` is supposed to be reinvoked as an actual generator nor am I sure whether this is happening on my system.

**Comment by someplaceguy** [comment]
Actually, I think it doesn't matter whether `systemd-fstab-generator` is being invoked as a generator, because `/run/wrappers` is not part of `/etc/fstab` so `systemd-fstab-generator` would never add a dependency between the `sysroot-run-wrappers.mount` unit and the `initrd-fs.target` unit.

So I guess this is why `/sysroot/run/wrappers` is being unmounted?

Or is there any other generator that is supposed to create this dependency? With this amount of obfuscation in systemd's implementation it's hard to tell... :)

**Comment by someplaceguy** [comment]
Actually, the same issue with `/sysroot/run/wrappers` exists with `/sysroot/nix/store`.

So now I have a system that has booted with systemd in initrd but `/nix/store` has not been properly bind mounted in read-only mode.

Are you able to run `touch /nix/store/test` as root without any errors, or does it fail with a `Read-only filesystem` error?

That command should fail, it should not succeed.

**Comment by someplaceguy** [comment]
Another question: do you have `/run/wrappers` and `/nix/store` in your `/etc/fstab`?

If not, why would you expect `systemd-fstab-generator` to add these filesystems as dependencies of `initrd-fs.target`?

**Comment by ElvishJerricco** [comment]
@someplaceguy `IgnoreOnIsolate` defaults to true for mount units. They shouldn't be stopped by an `isolate` command.

**Comment by misuzu** [comment]
Just experienced this issue after updating to b12803b6d90e2e583429bb79b859ca53c348b39a

**Comment by someplaceguy** [comment]
For the sake of anyone who isn't subscribed to #245089, I have posted some debug data there that could be helpful in diagnosing this issue as well, since I think they are related and may even have the same root cause (as both of them are caused by systemd unmounting filesystems in the exact same part of the boot process).

**Comment by someplaceguy** [comment]
@ElvishJerricco I'm finally pretty sure that I've found the root cause of this issue. I'm finally posting this comment from a NixOS system that has booted correctly with systemd in initrd for the first time since I've been trying. That is, `/run/wrappers` and `/nix/store` have now been properly mounted during boot, finally.

I don't think I could've figured this out with looking at the graph of systemd units and their dependencies/requirements, containing all the units in the initrd that are relevant.

The graph of dependencies I'm referring to is similar to the ones I posted in #245089. However, I've made the following improvements:

1. Instead of obtaining the graph from the emergency shell after `initrd-switch-root.service` has failed (due to the problem in #245089), I've obtained it from the emergency shell that started due to changing `initrd-cleanup.service` so that it contains `ExecStart=/bin/false` instead of `ExecStart=systemctl --no-block isolate ...`.

This means that I've ensured that `initrd-cleanup.service` always fails and never calls `systemd isolate`. This also allows me to log into the emergency shell and get the graph of dependencies at just the right point in the boot process.

In other words, this is a great improvement because it means that I've obtained the graph just before systemd was going to misbehave, which means that the filesystems are still mounted and we can inspect their dependencies properly to see which dependencies/requirements could cause them to be unmounted by the `systemctl isolate` command.

2. I've removed a lot of irrelevant clutter, by not only filtering out the edges representing ordering between units (it's clear that the ordering is not actually the problem, the problem is that the mount units are being stopped), but I've also filtered out other irrelevant units like all those temporary mountpoints I've created so that I could run `systemd-analyze` to obtain the graph of dependencies, all the `shutdown.target` conflict dependencies (it's clear that `shutdown.target` is not the problem, as we are booting rather than shutting down), all the emergency shell related dependencies (not really relevant either) and also all the `umount.target` conflicts.

The `umount.target` conflicts I'm referring to is the `Conflicts=umount.target` clause that each mount unit seems to "invisibly" contain (I say invisibly because I can't see it with e.g. `systemctl cat run-wrappers.mount`, but it can be seen with `systemd-analyze dot | grep run-wrappers.mount`).

Anyway, I think these conflicts are not really relevant because AFAIU they only exist to ensure that if `umount.target` is started, then all the mount units are stopped (i.e. all the filesystems are unmounted). 

This may appear to be suspiciously related to the problem I'm experiencing in this issue, because all (or almost all) the filesystems are in fact being unmounted, but I don't think it's related because `umount.target` has no other dependency except for these conflicts clauses. That said, besides the mount units, it does also conflict with `systemd-cryptsetup@root.service`, but I don't think that's problematic (?)

Anyway, these edges were also cluttering the graph visualization and I don't think they are relevant.

So here's the new graph visualization I've obtained after all these improvements:

[deps-stripped.pdf](https://github.com/NixOS/nixpkgs/files/12167257/deps-stripped.pdf)

Now, even with this improved visualization, it's not immediately clear what the problem is. But after some analysis I think it's pretty clear what is going on.

Here's what I think is happening during the boot process, in order:

1. First of all, I think everything works perfectly fine until we reach `initrd-cleanup.service`.
2. `initrd-cleanup.service`, as mentioned multiple times, starts the `systemctl --no-block isolate initrd-switch-root.target` command. So far so good.
3. This means that systemd was ordered to do the following:
   1. start `initrd-switch-root.target`
   2. stop everything that is not required or wanted by `initrd-switch-root.target`, or the requirements/wants of its requirements/wants, etc (recursively).
4. (We are still good up to this point)
5. `initrd-switch-root.target` "wants" `initrd-udevadm-cleanup-db.service` to start, so systemd starts making preparations to start it. This is normal and has been part of the upstream systemd boot process for at least 7 years.
6. `initrd-udevadm-cleanup-db.service` conflicts with `systemd-udev-settle.service`, so it cannot start until the latter is stopped, which means systemd starts making preparations to stop `systemd-udev-settle.service`. This is also normal and has been part of the upstream systemd boot process for at least 4 years.
7. However, `systemd-udev-settle.service` cannot stop immediately because `zfs-import-rpool.service` requires it. This is when things start to go wrong. Since systemd is making preparations to stop `systemd-udev-settle.service`, it now decides to stop `zfs-import-rpool.service`, because the former cannot stop without the latter stopping first.
8. `zfs-import-rpool.service` is a required dependency of `sysroot.mount`, which means that in order to stop `zfs-import-rpool.service`, systemd has to unmount `/sysroot` first. But of course, in order to unmount `/sysroot`, systemd has to unmount all the filesystems inside `/sysroot` first, including `/sysroot/tmp/wrappers` and `/sysroot/nix/store`.

So basically, it appears that stopping `systemd-udev-settle.service` is part of the normal, upstream systemd-in-initrd boot process, but since the ZFS root pool import service "requires" this udev service, then the service that imports the root pool is stopped as well, and this in turn causes all the mount units to be stopped as well (i.e. it causes the filesystems to  be mounted).

So if I'm diagnosing this issue correctly, it means that #245089 has exactly the same root cause as this one, and it also appears that this issue is 1) NixOS-specific, 2) only triggered if you enable both the systemd-in-initrd option, and also 3) only triggered if you use a ZFS root filesystem (well, similar issues could happen as well if you have some non-root ZFS filesystems, like `/nix/store`).

To test whether my theory was right, I applied this simple one-line patch to nixpkgs:
```diff
diff --git a/nixos/modules/tasks/filesystems/zfs.nix b/nixos/modules/tasks/filesystems/zfs.nix
index 1ab277a9907e..9533bdd41468 100644
--- a/nixos/modules/tasks/filesystems/zfs.nix
+++ b/nixos/modules/tasks/filesystems/zfs.nix
@@ -113,7 +113,7 @@ let
       # we need systemd-udev-settle to ensure devices are available
       # In the future, hopefully someone will complete this:
       # https://github.com/zfsonlinux/zfs/pull/4943
-      requires = [ "systemd-udev-settle.service" ];
+      wants = [ "systemd-udev-settle.service" ];
       after = [
         "systemd-udev-settle.service"
         "systemd-modules-load.service
```

What this patch does is pretty obvious: it makes `systemd-udev-settle.service` a "wanted" dependency of the services that import ZFS pools rather than a "required" dependency.

This means that if possible, `systemd-udev-settle.service` will still be started before the ZFS pool import services, but it also means two other important things:

1. If `systemd-udev-settle.service` is stopped, there is no need to stop the ZFS pool import services. This means that the filesystems that depend on these ZFS import services will not be stopped as well (in a sort of chain reaction). This should fix the problem I was experiencing.
2. As I understand it, if `systemd-udev-settle.service` fails, then the ZFS pool import services will be started anyway. This is a side effect of changing "requires" -> "wants", but not necessarily what was intended to fix the problem. This may or may not be a problem (I really don't know).

So, as you've probably figured out, I'm not entirely sure that my "fix" is correct, as there are other possible ways to fix the same issue. For example, as you can see from the tree of dependencies, an alternative way of fixing the problem would be to make `sysroot.mount` "want" `zfs-import-rpool.service` (instead of "require"), but this could have its own problems as well, as `zfs-import-rpool.service` and `zfs.target` would then be stopped before switching the root.

So I'm hoping someone like @ElvishJerricco could step in and see whether my fix is enough or if there's a better way to fix it.

Anyway, I hope this is it, finally, and if so, thanks for everything and I'm really sorry for all the noise I've caused while trying to debug these two issues (including all my extremely long comments, many of which about irrelevant points)...

**Comment by dcarosone** [comment]
I too just hit this issue, on an upgrade between 12303c65 and ef99fa5c, as noted in the above Discourse link.

**Comment by dcarosone** [comment]
> EDIT: after reboot everything works fine. Not sure what actually happened the first time.

Based on this comment, I decided to try rebooting the problematic generation again.

Attempt 1: same as the first time, an apparently normal boot, but no wrappers, sudo, firejail etc.
Attempt 2: systemd went to emergency mode failing to switch root. I hit enter to let it continue, and it did, system came up, and everything is fine (this is what I'm running now).

So there's definitely something racy involved.

I'm going to look at differences between them.  I don't have any custom changes to nixpkgs, and nothing particularly unusual in my config. Can also share logs if that's useful, but I'll see what else I can spot first.

**Comment by dcarosone** [comment]
Interesting:

```
diff -r -u system-16[89]-link/etc/systemd 
…
Only in system-169-link/etc/systemd/system: boot.mount.requires
Only in system-168-link/etc/systemd/system: boot.mount.wants
```

Both are a directory with just the zfs import service linked

```
❯ ls -Flas system-16[89]-link/etc/systemd/system/boot.mount.*
system-168-link/etc/systemd/system/boot.mount.wants:
total 91
 9 dr-xr-xr-x  2 root root   3 Jan  1  1970 ./
81 dr-xr-xr-x 76 root root 393 Jan  1  1970 ../
 1 lrwxrwxrwx  1 root root  27 Jan  1  1970 zfs-import-bpool.service -> ../zfs-import-bpool.service

system-169-link/etc/systemd/system/boot.mount.requires:
total 91
 9 dr-xr-xr-x  2 root root   3 Jan  1  1970 ./
81 dr-xr-xr-x 77 root root 394 Jan  1  1970 ../
 1 lrwxrwxrwx  1 root root  27 Jan  1  1970 zfs-import-bpool.service -> ../zfs-import-bpool.service
 ```

So, @someplaceguy it looks like your workaround is basically undoing a change between these revisions (which I haven't yet looked to track down)

**Comment by dcarosone** [comment]
```
systemd: 253.5 → 253.6, +201.8 KiB
```

**Comment by dcarosone** [comment]
Some other relevant diffs in systemd generated config:

```diff
diff -r -u system-168-link/etc/systemd/system/zfs-import.target.d/overrides.conf system-169-link/etc/systemd/system/zfs-import.target.d/overrides.conf
--- system-168-link/etc/systemd/system/zfs-import.target.d/overrides.conf       1970-01-01 10:00:01.000000000 +1000
+++ system-169-link/etc/systemd/system/zfs-import.target.d/overrides.conf       1970-01-01 10:00:01.000000000 +1000
@@ -1,4 +1,2 @@
 [Unit]
-After=zfs-import-bpool.service
-Requires=zfs-import-bpool.service

Only in system-169-link/etc/systemd/system: zfs-import.target.requires

diff -r -u system-168-link/etc/systemd/system/zfs-import-bpool.service system-169-link/etc/systemd/system/zfs-import-bpool.service
--- system-168-link/etc/systemd/system/zfs-import-bpool.service 1970-01-01 10:00:01.000000000 +1000
+++ system-169-link/etc/systemd/system/zfs-import-bpool.service 1970-01-01 10:00:01.000000000 +1000
@@ -1,19 +1,19 @@
 [Unit]
 After=systemd-udev-settle.service systemd-modules-load.service systemd-ask-password-console.service
-Before=boot.mount local-fs.target
+Before=boot.mount zfs-import.target
 DefaultDependencies=no
 Description=Import ZFS pool "bpool"
 Requires=systemd-udev-settle.service
```

At this point it is probably better to go looking for the source of these changes

**Comment by dcarosone** [comment]
I'm leaning more towards suspecting @someplaceguy's change being the correct fix, and basically being a missed part of the other change.

```text
❯ systemctl show systemd-udev-settle.service | egrep -i '(require|want|depend|before|after)'
RemainAfterExit=yes
Requires=system.slice
Wants=systemd-udevd.service
RequiredBy=zfs-import-bpool.service
Before=zfs-mount.service sysinit.target zfs-import-bpool.service
After=systemd-udev-trigger.service systemd-journald.socket system.slice
DefaultDependencies=no
```

Note that settle only 'wants', not 'requires', the udev service itself.

This system happens to be one where I haven't yet migrated the partitioning away from grub+bpool to systemd-boot+ESP-only, and I note that there's a bunch of stuff in here related to bpool.  I'm going to build an update for another system that has been migrated, and see how these dependencies fall out in that case.  It might explain why @ElvishJerricco hasn't seen this problem.

**Comment by dcarosone** [comment]
Yeah, without the bpool service, there's no chain of 'required' all the way from settle to `local-fs.target`, and that getting killed then cascades to `run-wrappers.mount`or others also getting killed.

**Comment by dcarosone** [comment]
Oh, I should add, though I think it's probably obvious: I'm looking at the post-pivotroot stage2 systemd configuration. I think these two issues are the same bug, one in each stage, and affecting different mounts. I don't really know why/how the settle service might be getting stopped in stage 2, though.  I definitely feel like there are probably some other mounts affected/missing too, this one just happens to be easy to notice.

And, when the problem is viewed as a chain of 'requires' all the way from 'settle' to 'local-fs', it's clear why this is the missing part of the other fixes in the recent commit: we need to break the chain, but it belongs at the early end not the late end.
